modules/service/src/main/resources/db/migration/V0843__program_reference.sql

```diff
@@ -0,0 +1,90 @@
+-- Add a semester column to program.  It is nullable but there will be a
+-- constraint that checks that it is set when the proposal status is not
+-- 'not_submitted'.
+
+CREATE DOMAIN d_semester AS varchar
+  CHECK (VALUE ~ '^[0-9][0-9][0-9][0-9]+[AB]$');
+COMMENT ON DOMAIN d_semester IS 'Formatted Semester';
+
+ALTER TABLE t_program
+  ADD COLUMN c_semester d_semester;
+
+-- Default semester for existing submitted programs.
+UPDATE t_program
+  SET
+    c_semester =
+        CONCAT(
+            LPAD(EXTRACT(YEAR FROM CURRENT_TIMESTAMP AT TIME ZONE 'UTC')::text, 4, '0'),
+            CASE
+                WHEN EXTRACT(MONTH FROM CURRENT_TIMESTAMP AT TIME ZONE 'UTC') BETWEEN 2 AND 7 THEN 'A'
+                ELSE 'B'
+            END
+        )::d_semester
+  WHERE
+    c_proposal_status <> 'not_submitted';
+
+-- A function that obtains the next index for a semester
+CREATE OR REPLACE FUNCTION next_semester_index(semester d_semester)
+RETURNS INT
+AS $$
+DECLARE
+    sequence_name VARCHAR := CONCAT('s_', semester);
+BEGIN
+    EXECUTE 'CREATE SEQUENCE IF NOT EXISTS ' || sequence_name;
+    RETURN nextval(sequence_name)::INT;
+END;
+$$ LANGUAGE plpgsql;
+
```

Comment:
Oh, I like the idea of creating the sequences lazily. I would not have thought of this.